### PR TITLE
Economy: Update Economy Stats

### PR DESCRIPTION
**Pull request by Lord-Haji**

### CODE REVIEW

**Comment by HoeenCoder** [comment]
What we have is fine. I literally just merged it an hour ago

**Comment by HoeenCoder** [comment]
nvrm i didnt read it properly :x

File: spacialgaze-plugins/economy.js [hunk 1]
```diff
@@ -325,13 +325,17 @@ exports.commands = {
...
 		if (!this.runBroadcast()) return;
-		const users = Object.keys(Db.currency.object());
-		const total = users.reduce(function (acc, cur) {
-			return acc + Db.currency.get(cur);
-		}, 0);
+		const users = Db.currency.keys();
+		let total = 0;
+		for (let i = 0; i < users.length; i++) {
+			total += Economy.readMoney(users[i]);
+			console.log (Economy.readMoney(users[i]));
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js]: Dont need to log this

File: spacialgaze-plugins/economy.js [hunk 2]
```diff
@@ -325,13 +325,17 @@ exports.commands = {
 	stardust: 'economystats',
 	economystats: function (target, room, user) {
 		if (!this.runBroadcast()) return;
-		const users = Object.keys(Db.currency.object());
-		const total = users.reduce(function (acc, cur) {
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js:329]: Whats wrong with array#reduce?

File: spacialgaze-plugins/economy.js [hunk 3]
```diff
@@ -325,13 +325,17 @@ exports.commands = {
...
+			total += Economy.readMoney(users[i]);
+			console.log (Economy.readMoney(users[i]));
+		}
 		let average = Math.floor(total / users.length) || 0;
-		let output = "There " + (total > 1 ? "are " : "is ") + total + (total > 1 ? currencyPlural : currencyName) + " circulating in the economy. ";
-		output += "The average user has " + average + (average > 1 ? currencyPlural : currencyName) + ".";
+		let output = "There " + (total > 1 ? "are " : "is ") + "<strong>" + total + " " + (total > 1 ? currencyPlural : currencyName) + "</strong> circulating in the economy. <br />";
+		output += "The average user has <strong>" + average + " " + (average > 1 ? currencyPlural : currencyName) + "</strong>.<br />";
+		let hasAStardust = Math.round(total / users.length);
+		output += "At least <strong>" + Math.round(total / users.length) + "</strong> users has a <strong>" + currencyName + "</strong>";
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js]: Wouldnt the number of users who have 1 stardust _not_ be the average?

Comment (DeathlyPlays) [reply spacialgaze-plugins/economy.js]: It's supposed to add EVERY user lol

Comment (DeathlyPlays) [reply spacialgaze-plugins/economy.js]: Well users since this command is introduced

Comment (HoeenCoder) [reply spacialgaze-plugins/economy.js]: Some users could be recorded but have 0 dust, tbh just `filter()` the array

Comment (ghost) [reply spacialgaze-plugins/economy.js]: ^ people who come but are broke should be added but not everyone came to sg

File: spacialgaze-plugins/economy.js (hunk 2 above)
Comment (Lord-Haji) [reply spacialgaze-plugins/economy.js:329]: for loop has better performance

**Comment by Lord-Haji** [comment]
try something funny no idea whether it will work :P

File: spacialgaze-plugins/economy.js [hunk 4]
```diff
@@ -335,8 +335,8 @@ exports.commands = {
 	stardust: 'economystats',
 	economystats: function (target, room, user) {
 		if (!this.runBroadcast()) return;
-		let average = Math.floor(total / users.length) || 0;
-		let output = "There " + (total > 1 ? "are " : "is ") + "<strong>" + economyStat()[1] + " " + (economyStat()[1] > 1 ? currencyPlural : currencyName) + "</strong> circulating in the economy. <br />";
+		let average = Math.floor(economyStat()[1] / users.length) || 0;
```
Comment (ghost) [inline spacialgaze-plugins/economy.js]: Users isn't defined. I think it's supposed to be user???

Comment (Lord-Haji) [reply spacialgaze-plugins/economy.js]: nope

Comment (Lord-Haji) [reply spacialgaze-plugins/economy.js]: but yeah i did something wrong

Comment (Lord-Haji) [reply spacialgaze-plugins/economy.js]: I mean i fixed locally but forgot to push

**Comment by Lord-Haji** [comment]
bump @HoeenCoder

**Comment by Lord-Haji** [comment]
Rebump @HoeenCoder

**Comment by HoeenCoder** [comment]
Very busy in RL atm, havent been checking

File: spacialgaze-plugins/economy.js [hunk 5]
```diff
@@ -114,6 +115,15 @@ function rankLadder(title, type, array, prop, group) {
 	return ladderTitle + tableTop + tableRows + tableBottom;
 }
 
+function economyStat() {
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js:118]: Do we need this function? Were only calling it in one place tbh, might just want to implement it in the comand.

File: spacialgaze-plugins/economy.js [hunk 6]
```diff
@@ -325,13 +335,10 @@ exports.commands = {
...
 	economystats: function (target, room, user) {
 		if (!this.runBroadcast()) return;
-		const users = Object.keys(Db.currency.object());
-		const total = users.reduce(function (acc, cur) {
-			return acc + Db.currency.get(cur);
-		}, 0);
-		let average = Math.floor(total / users.length) || 0;
-		let output = "There " + (total > 1 ? "are " : "is ") + total + (total > 1 ? currencyPlural : currencyName) + " circulating in the economy. ";
-		output += "The average user has " + average + (average > 1 ? currencyPlural : currencyName) + ".";
+		let average = Math.floor(economyStat()[1] / economyStat()[2]) || 0;
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js:338]: Dont call economyStat multiple times, assign the results to a variable.

File: spacialgaze-plugins/economy.js [hunk 7]
```diff
@@ -325,13 +335,10 @@ exports.commands = {
...
+		let output = "There " + (economyStat()[1] > 1 ? "are " : "is ") + "<strong>" + economyStat()[1] + " " + (economyStat()[1] > 1 ? currencyPlural : currencyName) + "</strong> circulating in the economy. <br />";
+		output += "The average user has <strong>" + average + " " + (average > 1 ? currencyPlural : currencyName) + "</strong>.<br />";
+		output += "<strong>" + economyStat()[0] + "</strong> users has a " + currencyName;
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js:341]: ...`users have at least 1 " + currencyName;`

File: spacialgaze-plugins/economy.js [hunk 8]
```diff
@@ -114,6 +115,15 @@ function rankLadder(title, type, array, prop, group) {
 	return ladderTitle + tableTop + tableRows + tableBottom;
 }
 
+function economyStat() {
+	let users = Db.currency.keys(), hasOne, total = 0;
+	for (let i = 0; i < users.length; i++) {
+		if (Economy.readMoney(users[i]) === 1) hasOne++;
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js:121]: \>= 1 not === 1

File: spacialgaze-plugins/economy.js [hunk 9]
```diff
@@ -114,6 +115,15 @@ function rankLadder(title, type, array, prop, group) {
...
+		total += Economy.readMoney(users[i]);
```
Comment (HoeenCoder) [inline spacialgaze-plugins/economy.js:122]: Incase of undefined or another weird non-number value, make this `total += Economy.readMoney(users[i]) || 0;`

File: spacialgaze-plugins/economy.js (hunk 8 above)
Comment (Lord-Haji) [reply spacialgaze-plugins/economy.js:121]: Right

File: spacialgaze-plugins/economy.js (hunk 5 above)
Comment (Lord-Haji) [reply spacialgaze-plugins/economy.js:118]: True

**Pull request closed without merging**